runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+        final ProcessContext baseContext,
+        final WindowedValue<InputT> element,
+        final TrackerT tracker,
+        final RestrictionT[] residualRestrictionHolder,
+        final WindowingInternals<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
+            internals) {
+      return fn.new ProcessContext() {
+        // Commit at least once every 10k output records.  This keeps the watermark advancing
+        // smoothly, and ensures that not too much work will have to be reprocessed in the event of
+        // a crash.
+        // TODO: Also commit at least once every N seconds (runner-specific parameter).
+        private static final int MAX_OUTPUTS_PER_BUNDLE = 10000;
+
+        private int numOutputs = 0;
+
+        public InputT element() {
+          return element.getValue();
+        }
+
+        public Instant timestamp() {
+          return element.getTimestamp();
+        }
+
+        public PaneInfo pane() {
+          return element.getPane();
+        }
+
+        public void output(OutputT output) {
+          internals.outputWindowedValue(
+              output, element.getTimestamp(), element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        public void outputWithTimestamp(OutputT output, Instant timestamp) {
+          internals.outputWindowedValue(output, timestamp, element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        private void noteOutput() {
+          if (++numOutputs > MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
+            residualRestrictionHolder[0] = tracker.checkpoint();
+          }
+        }
+
+        public <T> T sideInput(PCollectionView<T> view) {
+          return baseContext.sideInput(view);
+        }
+
+        public PipelineOptions getPipelineOptions() {
+          return baseContext.getPipelineOptions();
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
+          baseContext.sideOutputWithTimestamp(tag, output, element.getTimestamp());
+        }
+
+        public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
+          baseContext.sideOutputWithTimestamp(tag, output, timestamp);
+        }
+      };
+    }
+
+    /** Creates an {@link DoFn.ExtraContextFactory} that provides just the given tracker. */
+    private DoFn.ExtraContextFactory<InputT, OutputT> wrapTracker(final TrackerT tracker) {
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          throw new UnsupportedOperationException("Splittable DoFn's don't support extra context");
```

Comment:
And in fact this shouldn't be reachable, since the user wasn't able to ask for any of these, correct?